### PR TITLE
Gan dev pyt

### PR DESCRIPTION
**Pull request by gallupBenRyan**
Note: `src/gan/discriminator/` files can be ignored
           `src/gan/generator/` files can be ignored, because these are going away.

Just `src/gan` and `src/gan/gen_pyt` need review (and others in other directories if they are included, like `src/rdf_graph/rdf_parse` etc.)

### CODE REVIEW

**Comment by BoltGallup** [comment]
@gallupBenRyan I went over the `rdf_parse` and `map_queries` and they look good to go. Are there any parts of the GAN that you think that I should take a look?

File: src/gan/config.cfg [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+SEED = 88
+BATCH_SIZE = 64
+TOTAL_EPOCHS = 200 
+GENERATED_NUM = 100
+VOCAB_SIZE = 10
+SEQUENCE_LEN = 20
+
+data = 
+data_dir = data
+glove_dir = /Users/ben_ryan/Documents/git/multivac/models
```
Comment (mhoover-gallup) [inline src/gan/config.cfg]: Change to a relative path.

File: src/gan/config.cfg [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+seq2tree_sample_file = model.sample
+seq2tree_id_file = test.id.txt
+seq2tree_rareword_map = None
+seq2seq_decode_file = 
+seq2seq_ref_file = 
+is_nbest = False
+
+
+[DISCRIMINATOR]
+data = discriminator/data/multivac
```
Comment (mhoover-gallup) [inline src/gan/config.cfg]: let's ensure folder structure is the same locally and on AWS.

File: src/gan/data_utils.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
+#!usr/bin/env/python
+import numpy as np
+
+import torch
+import torch.utils.data as data
+
+from multivac.src.gan.generator.dataset import DataEntry, Vocab, DataSet
```
Comment (mhoover-gallup) [inline src/gan/data_utils.py]: @mitchbregs will these imports work within Docker or do they need to change?

File: src/gan/gen_pyt/asdl/__init__.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# import six
```
Comment (mhoover-gallup) [inline src/gan/gen_pyt/asdl/__init__.py]: cool to have all this commented out? is the `__init__.py` even needed?

File: src/gan/gen_pyt/asdl/asdl.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
+# coding=utf-8
```
Comment (mhoover-gallup) [inline src/gan/gen_pyt/asdl/asdl.py]: this does nothing. should be `# -*- coding: utf-8 -*-`

File: src/gan/gen_pyt/asdl/asdl_ast.py [hunk 6]
```diff
@@ -0,0 +1,213 @@
+# coding=utf-8
+
+from io import StringIO
+
+from multivac.src.gan.gen_pyt.asdl import *
```
Comment (mhoover-gallup) [inline src/gan/gen_pyt/asdl/asdl_ast.py]: let's ensure -- for any file -- that repo-based imports like this will work as-is with docker.

File: src/gan/gen_pyt/asdl/asdl_ast.py [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+            assert len(realized_fields) == len(self.production.fields)
+
+            for field in realized_fields:
+                self.add_child(field)
+        else:
+            for field in self.production.fields:
+                self.add_child(RealizedField(field))
+
+    def add_child(self, realized_field):
+        # if isinstance(realized_field.value, AbstractSyntaxTree):
```
Comment (mhoover-gallup) [inline src/gan/gen_pyt/asdl/asdl_ast.py]: Delete?

File: src/gan/gen_pyt/asdl/asdl_ast.py [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+        #     realized_field.value.parent = self
+        self.fields.append(realized_field)
+        realized_field.parent_node = self
```
Comment (mhoover-gallup) [inline src/gan/gen_pyt/asdl/asdl_ast.py]: Not sure I understand what's going on here -- `realized_field.parent_node` is going to inherit the entire class?

File: src/gan/gen_pyt/asdl/hypothesis.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
+# coding=utf-8
+
+from multivac.src.gan.gen_pyt.asdl import *
```
Comment (mhoover-gallup) [inline src/gan/gen_pyt/asdl/hypothesis.py]: flagging the imports for checking.

File: src/gan/gen_pyt/asdl/lang/eng/eng_asdl_helper.py [hunk 10]
```diff
@@ -0,0 +1,75 @@
+# coding=utf-8
+
+from multivac.src.gan.gen_pyt.asdl.asdl_ast import RealizedField, AbstractSyntaxTree
```
Comment (mhoover-gallup) [inline src/gan/gen_pyt/asdl/lang/eng/eng_asdl_helper.py]: import flag

File: src/gan/gen_pyt/astnode.py [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+    ])
+
+    t2 = ASTNode('root', children=[
+        ASTNode(str, 'a1_label', children=[ASTNode(int, children=[ASTNode('a21', value=123)]),
+                                           ASTNode(ast.NodeTransformer, children=[ASTNode('a21', value='hahaha')])]
+                ),
+        ASTNode('a2', children=[ASTNode('a21', value='asdf')])
+    ])
+
+    print((t1 == t2))
```
Comment (mhoover-gallup) [inline src/gan/gen_pyt/astnode.py]: needed? this line but everything below as well.

File: src/gan/gen_pyt/common/utils.py [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+        self.func = func
+
+    def __get__(self, obj, cls):
+        if obj is None:
+            return self
+        value = obj.__dict__[self.func.__name__] = self.func(obj)
+        return value
+
+
+def init_arg_parser():
```
Comment (mhoover-gallup) [inline src/gan/gen_pyt/common/utils.py]: I assume all of this was coded by someone else. if not, it might make more sense to turn this to a config.

File: src/gan/gen_pyt/components/__init__.py [hunk 13]
```diff
@@ -0,0 +1,8 @@
+# import six
```
Comment (mhoover-gallup) [inline src/gan/gen_pyt/components/__init__.py]: Same comments as other `__init__.py` file -- needed?

File: src/gan/gen_pyt/components/action_info.py [hunk 14]
```diff
@@ -0,0 +1,64 @@
+# coding=utf-8
+from multivac.src.gan.gen_pyt.asdl.hypothesis import Hypothesis
```
Comment (mhoover-gallup) [inline src/gan/gen_pyt/components/action_info.py]: import reminder.

File: src/gan/gen_pyt/components/dataset.py [hunk 15]
```diff
@@ -0,0 +1,227 @@
+# coding=utf-8
+from collections import OrderedDict
+
+import torch
+import numpy as np
+import pickle
+
+from torch.autograd import Variable
+
+from multivac.src.gan.gen_pyt.asdl.transition_system import ApplyRuleAction, ReduceAction
```
Comment (mhoover-gallup) [inline src/gan/gen_pyt/components/dataset.py]: import reminder

File: src/gan/gen_pyt/components/dataset.py [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+
+    def __len__(self):
+        return len(self.examples)
+
+    def get_frontier_field_idx(self, t):
+        ids = []
+        for e in self.examples:
+            if t < len(e.tgt_actions):
+                ids.append(self.grammar.field2id[e.tgt_actions[t].frontier_field])
+                # assert self.grammar.id2field[ids[-1]] == e.tgt_actions[t].frontier_field
```
Comment (mhoover-gallup) [inline src/gan/gen_pyt/components/dataset.py]: Use or delete.

File: src/gan/gen_pyt/components/dataset.py [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+                ids.append(0)
+
+        return Variable(torch.cuda.LongTensor(ids)) if self.cuda else Variable(torch.LongTensor(ids))
+
+    def get_frontier_prod_idx(self, t):
+        ids = []
+        for e in self.examples:
+            if t < len(e.tgt_actions):
+                ids.append(self.grammar.prod2id[e.tgt_actions[t].frontier_prod])
+                # assert self.grammar.id2prod[ids[-1]] == e.tgt_actions[t].frontier_prod
```
Comment (mhoover-gallup) [inline src/gan/gen_pyt/components/dataset.py]: Use or delete

File: src/gan/gen_pyt/components/dataset.py [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+                ids.append(0)
+
+        return Variable(torch.cuda.LongTensor(ids)) if self.cuda else Variable(torch.LongTensor(ids))
+
+    def get_frontier_field_type_idx(self, t):
+        ids = []
+        for e in self.examples:
+            if t < len(e.tgt_actions):
+                ids.append(self.grammar.type2id[e.tgt_actions[t].frontier_field.type])
+                # assert self.grammar.id2type[ids[-1]] == e.tgt_actions[t].frontier_field.type
```
Comment (mhoover-gallup) [inline src/gan/gen_pyt/components/dataset.py]: use or delete

File: src/gan/gen_pyt/datasets/english/dataset.py [hunk 19]
```diff
@@ -0,0 +1,247 @@
...
+            # Remove non-alphabetic characters at the start of the string
+            query = query.strip()
+            query = re.sub(r"“|”", "\"", query)
+            query = re.sub(r"‘|’", "\'", query)
+            query = re.sub(r"`", "\'", query)
+            query = query.strip("\"")
+            query = query.strip("\'")
+            query = query[query.index(re.split(r"\"", query)[-1]):]
+            query = query[query.index(re.split(r"NumericCitation", query, re.IGNORECASE)[-1]):]
+            query = query[query.index(re.split(r"[\.\!\?]\s+", query)[-1]):]
```
Comment (mhoover-gallup) [inline src/gan/gen_pyt/datasets/english/dataset.py]: Some of these could benefit from a comment above talking about what it's removing. figuring it out from the pattern is a little tough at a quick glance.

File: src/gan/gen_pyt/datasets/english/eng_dataset.py [hunk 20]
```diff
@@ -0,0 +1,373 @@
...
+                    tok_src_idx = query_tokens.index(terminal_token)
+                except ValueError:
+                    pass
+
+                d = {'literal': terminal_token, 'rule': rule, 
+                     'parent_rule': parent_rule, 'parent_t': parent_t}
+
+                # cannot copy, only generation
+                # could be unk!
+                # import pdb; pdb.set_trace()
```
Comment (mhoover-gallup) [inline src/gan/gen_pyt/datasets/english/eng_dataset.py]: please remove, even if commented out.

File: src/gan/gen_pyt/gen_config.cfg [hunk 21]
```diff
@@ -0,0 +1,98 @@
+
+#### General configuration ####
```
Comment (mhoover-gallup) [inline src/gan/gen_pyt/gen_config.cfg]: this feels like another config that was reviewed earlier in the PR... are they different in substantive ways?

File: src/gan/config.cfg (hunk 2 above)
Comment (gallupBenRyan) [reply src/gan/config.cfg]: This should be a local path from `config.cfg` which sits in `/src/gan` so should be fine. Also, changed the above mentioned path to be a local one as requested.

File: src/gan/gen_pyt/asdl/__init__.py (hunk 4 above)
Comment (gallupBenRyan) [reply src/gan/gen_pyt/asdl/__init__.py]: We need the file, but I don't believe it needs to have anything in it. I haven't confirmed that yet, though, so I commented it for now before fully deleting.

File: src/gan/gen_pyt/asdl/asdl_ast.py (hunk 8 above)
Comment (gallupBenRyan) [reply src/gan/gen_pyt/asdl/asdl_ast.py]: Each node in the AbstractSyntaxTree is basically defined as a sub-tree, with child nodes as "RealizedFields" - each of which is an AbstractSyntaxTree itself. Each AST remembers its place in the tree with a `parent_node` property that links back to the parent node/AST.

File: src/gan/gen_pyt/common/utils.py (hunk 12 above)
Comment (gallupBenRyan) [reply src/gan/gen_pyt/common/utils.py]: Actually in process implementing that as we speak! Diana had a question about the file that calls this which highlighted the issue for me, too.

File: src/gan/gen_pyt/gen_config.cfg (hunk 21 above)
Comment (gallupBenRyan) [reply src/gan/gen_pyt/gen_config.cfg]: This is specific to the new pytorch version of the generator, but based on this and another config/args parsing comment above, I'm merging the two cfg files and updating to take all GAN parameters from one `config.cfg` file.

File: src/gan/gen_pyt/model/utils.py [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+from torch.autograd import Variable
+
+from multivac.src.gan.gen_pyt.model.parser import Parser
+
+class GloveHelper(object):
+    def __init__(self, glove_file):
+        self.glove_file = glove_file
+        embeds = np.zeros((5000, 100), dtype='float32')
+        for i, (word, embed) in enumerate(self.embeddings):
+            if i == 5000: break
```
Comment (mhoover-gallup) [inline src/gan/gen_pyt/model/utils.py]: why only the first 5000? does that matter in any significant way? is it fine to hard-code this or should it be variable?

File: src/gan/querygan.py [hunk 23]
```diff
@@ -0,0 +1,335 @@
...
+    netG.build()
+
+    # Set up Discriminator component with given parameters
+
+    dargs['vocab_size'] = glove_vocab.size()
+    netD = QueryGAN_Discriminator(dargs)
+    trainer = disc_trainer(netD, glove_emb, glove_vocab, use_cuda)
+
+    # Set up Oracle component with given parameters
+    # ### This is super expensive memory wise. let's figure something else out
```
Comment (mhoover-gallup) [inline src/gan/querygan.py]: what's going on here? comment is valuable or no? do we have another way?

File: src/gan/querygan.py [hunk 24]
```diff
@@ -0,0 +1,335 @@
...
+                        help='Enable GPU training.')
+    parser.add_argument('-c', '--config', required=False, 
+                        help='Config file with updated parameters for generator;'
+                             'defaults to "config.cfg" in this directory '
+                             'otherwise.')
+
+    all_args = parser.parse_known_args()
+    args = vars(all_args[0])
+
+    i = 0
```
Comment (mhoover-gallup) [inline src/gan/querygan.py]: What's this about? Seems to have no meaning if the code below is commented out.

File: src/gan/querygan_pyt.py [hunk 25]
```diff
@@ -0,0 +1,367 @@
+#!usr/bin/env/python
```
Comment (mhoover-gallup) [inline src/gan/querygan_pyt.py]: Ultimately, what's the difference betwee n `querygan.py` and `querygan_pyt.py`? are both needed? can one be deleted?

File: src/gan/utilities/losses.py [hunk 26]
```diff
@@ -0,0 +1,46 @@
+#!usr/bin/env/python
```
Comment (mhoover-gallup) [inline src/gan/utilities/losses.py]: This file feels very similar to something already reviewed; are both needed?

File: src/rdf_graph/rdf_parse.py [hunk 27]
```diff
@@ -236,8 +237,9 @@ def get_rdfs(self, use_tokens=True, how='longest'):
                                           for t 
                                           in rdf[part]])
                         # result[rdf] = [x for x in result[rdf].values()]
```
Comment (mhoover-gallup) [inline src/rdf_graph/rdf_parse.py]: Remove?

File: src/gan/gen_pyt/model/neural_lm.py [hunk 28]
```diff
@@ -0,0 +1,97 @@
...
+                return [self.vocab.id2word[wid] for wid in sample_words[1:]]
+            else:
+                sample_words.append(x_t_wid)
+
+            h_tm1 = last_state, last_cell
+
+    @classmethod
+    def load(self, model_path, cuda=False):
+        params = torch.load(model_path, map_location=lambda storage, loc: storage)
+        model = LSTMLanguageModel(params['vocab'], *params['args'])
```
[the remark this answers is not in the document]
Comment (gallupBenRyan) [reply src/gan/gen_pyt/model/neural_lm.py]: Yeah, this should be fine. I removed a mechanism for updating parameters/settings in a previously trained model in this function. That doesn't affect the parameters already previously set.

File: src/gan/querygan_pyt.py (hunk 25 above)
Comment (gallupBenRyan) [reply src/gan/querygan_pyt.py]: Moving from querygan.py to the _pyt version as we speak. The non-pyt one used Theano for the generator system. I've got both right now because that move is still in transition.

File: src/gan/gen_pyt/model/utils.py (hunk 22 above)
Comment (gallupBenRyan) [reply src/gan/gen_pyt/model/utils.py]: Noting this to follow up on in an upcoming new dev branch, after we get this checkpoint PR merged. 5000 is the default vocab size, set/defined in a couple places. I want to open this up to be arbitrary size, but that will require changes in multiple places to keep everything aligned.

File: src/gan/utilities/losses.py (hunk 26 above)
Comment (gallupBenRyan) [reply src/gan/utilities/losses.py]: This is where we define the adversarial loss for the GAN system. I'm not sure what this resembles - potentially an earlier version somewhere?
